docs/BUILD_SYSTEM.md

```diff
@@ -0,0 +1,78 @@
+# The Service-Catalog Build System
+
+This document aims to overview the build system in this repository.
+
+## Composition
+
+This repository contains several Makefiles that enable the
+[GNU Make](https://www.gnu.org/software/make/manual/make.html) tool to build
+the code herein.
+
+Several components are built from this repository (catalog controller,
+example brokers, ...). One or more artifacts are constructed for each component
+(e.g. docker images, binaries). Generally speaking there is a single
+Makefile per component.
+
+## Structure
+
+There is a top-level [Makefile](./Makefile) that drives high-level actions
+(i.e. installing repository-wide Go dependencies) and recursively builds other
+targets in each of the directories listed in the DIRS variable.
+
+For example, typing make build at the top level will effectively run
+`make -C ${DIR}` build for each DIR in DIRS.
+
+The subdirectory Makefiles are designed to be used both by calling
+`make <target>` from top level (in which case `make <target>` in all
+subdirectories will be executed) but also as standalone Makefiles. This is
+to make it possible to run make in each individual subdirectory, and
+operate on subset of the code base.
+
+
+## Subdirectory Makefiles
+
+Each subdirectory that contains code that must be built into an artifact has
+a Makefile in it. Generally, these Makefiles are simple because they specify
+a few variables, and the remainder of the Makefile code is included from
+`hack/Makefile.mk` and `hack/Common.mk`.
+
+Below is a rough outline of one of these subdirectory `Makefile`s (eliding the
+copyright notice):
+
```

Comment:
I would either talk about the general flow of Makefile.mk and Common.mk w/o including snippets or leave it out all together. The problem with having snippets of code here is that its sure to get out of sync with the real Makefile and then it loses its value and could just confuse people even more.  Plus, I'm not sure I see the value of showing how we set a few env var & some "include"s w/o any real logic.